Repo Gardening: add task to gather support references in a comment

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

- Gather all support references in the original issue and all comments on the issue.
- Build a list of unique, correct support references from that list of HE-provided references.
- Post that list into a new comment.
- Update that list whenever the issue is updated, reopened, or a new issue comment is created.

**Notes**

- We do not want to update the comment when an issue comment is edited; doing so would stop someone from checking the boxes in that very comment, their changes would get reverted every time.
- We currently support a few support formats, regardless of case: xxx-chat, xxx-hc, xxx-zen, xxx-zd, xxx-zd-woothemes. We do not support references of the following format (yet): hc-xxx, zd-xxx.
- This allows us to tackle the problem #24947 meant to fix, with a different approach.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

Internal references:
- pciE2j-18N-p2
- pcLjiI-QJ-p2#comment-806

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

This will need to be merged before we can test this. You can see this in action in this issue:
https://github.com/jeherve/jetpack/issues/43

jeherve committed:
Repo Gardening: add task to gather support references in a comment

- Gather all support references in the original issue and all comments on the issue.
- Build a list of unique, correct support references from that list of HE-provided references.
- Post that list into a new comment.
- Update that list whenever the issue is updated, reopened, or a new issue comment is created.

**Notes**

- We do not want to update the comment when an issue comment is edited; doing so would stop someone from checking the boxes in that very comment, their changes would get reverted every time.
- We currently support a few support formats, regardless of case: xxx-chat, xxx-hc, xxx-zen, xxx-zd, xxx-zd-woothemes. We do not support references of the following format (yet): hc-xxx, zd-xxx.

Internal reference: pcLjiI-QJ-p2

jeherve committed:
changelog

jeherve committed:
Do not trigger task when comments are edited.

jeherve committed:
Also trigger action when an issue body is edited

jeherve committed:
fix owner reference

jeherve committed:
Remove tab indentation for reference

jeherve committed:
Fix list display

jeherve commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,165 @@
+	} );
+	ticketReferences.push( ...body.matchAll( referencesRegexP ) );
+
+	// Buid a first array with only the support IDs we've collected.
+	ticketReferences.map( reference => {
+		supportIds.push( reference[ 0 ] );
+	} );
+
+	// That array can still include duplicates, or references that are not formatted quite properly.
+	// Let's build a final array with unique and correct support IDs.
+	const correctedSupportIds = [];
+	supportIds.map( supportId => {
+		// xxx-zen and xxx-hc are the preferred formats for tickets and chats.
+		// xxx-zd and xxx-chat, as well as uppercase versions, are considered as alternate versions.
+		const wrongId = supportId.match( /([0-9]*)-(zd|chat)/i );
+		if ( wrongId ) {
+			const correctedId = `${ wrongId[ 1 ] }-${ wrongId[ 2 ] === 'zd' ? 'zen' : 'hc' }`;
+			if ( ! correctedSupportIds.includes( correctedId ) ) {
+				correctedSupportIds.push( correctedId );
+			}
+		} else if ( ! correctedSupportIds.includes( supportId.toLowerCase() ) ) {
+			correctedSupportIds.push( supportId.toLowerCase() );
+		}
+	} );
```
I wonder if there is a better, simpler way of doing this. I wanted to split things a bit to make them more readable, but I worry it's still too complex.

anomiex replied:
You could do
```
const correctedSupportIds = new Set();
```
then you can just `correctedSupportIds.add()` instead of having to guard it with `if ( ! correctedSupportIds.includes( correctedId ) )`.

Converting back to an array can be done with `[ ...correctedSupportIds ]` if you need to.

You might also switch from `.map()` to `.forEach()`, since you're not using the return value. Or you could so far as to do `for ( const supportId of supportIds )` instead, but I suspect readability of that versus `.forEach()` is a matter of taste.

jeherve replied:
It does simplify things a bit. 👍 done in 051ed1783e30eeb1363247b3c4f0a71c84893e72

jeherve committed:
Update docblock

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,165 @@
+	for await ( const response of octokit.paginate.iterator( octokit.rest.issues.listComments, {
+		owner,
+		repo,
+		issue_number: +number,
```
Let's avoid unnecessary pagination.
```suggestion
		issue_number: +number,
		per_page: 100,
```

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,165 @@
+	const referencesRegexP = /[0-9]*-(?:chat|hc|zen|zd)/gim;
+
+	debug( `gather-support-references: Getting references from comments.` );
+	for await ( const response of octokit.paginate.iterator( octokit.rest.issues.listComments, {
```
You're fetching the list of comments twice. It'd be better to fetch it just once then process it multiple times. Save the API hits.

jeherve replied:
That's a good point. Maybe that's also a good opportunity to extract this, since we already use it elsewhere. I took care of that in dce34a220fb01687d3d6672443d4ca09756f1e9c and 7d8127cd7d521b33207dcb71c0cc8c59a5abdb40

jeherve replied:
Hm, I think I broke things with `addMilestone` and `wpcomCommitReminder`; I'll take another look at this.

Edit: hm no, it does seem okay.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/ticket-reference-comment

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,165 @@
+		issue_number: +number,
+	} ) ) {
+		response.data.map( comment => {
+			ticketReferences.push( ...comment.body.matchAll( referencesRegexP ) );
```
Might be worth having the bot ignore its own comment. Otherwise if someone adds a wrong task, then deletes it, the bot would preserve it since it would still find it in its own comment.

jeherve replied:
That's a good point. Done in 59903a1a33e06f7150ddafab548eedaf8bab6146

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,165 @@
+	${ issueReferences
+		.map(
+			reference => `
+- [ ] ${ reference }`
```
So every time the bot updates its comment, it'll uncheck all the checkboxes. That doesn't seem like the behavior we want.

jeherve replied:
From what I know, HEs need the checkboxes so they can get back to the users once the issue is closed, and keep track of the users they've contacted.

With that in mind, and since the action is not triggered when issue comments are edited, I don't think it should get triggered once the issue has been closed.

That said, maybe I should add an additional safeguard anyway, so a new comment is created when comments are left once the issue is opened? I'm not sure that's the best way to go about this.

anomiex replied:
If someone leaves a new comment on the issue, even after it's closed, wouldn't that trigger another run?

Personally I'd just scan through the bot's comment to collect all the references that look like `[x] reference`, then use that to output the `[x]` again here.

jeherve replied:
I took a stab at this in df0d52c64c1f9f44924ab9d4213252330d768392 (and 20e0387c6a51e890d184975432a757c218110823)

jeherve committed:
avoid unnecessary pagination

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Extract comment fetching call to its own utility.

See https://github.com/Automattic/jetpack/pull/25066/files#r919122536

jeherve committed:
Only get all comments on issue once

See https://github.com/Automattic/jetpack/pull/25066/files#r919122536

anomiex committed:
Switch from array to Set to better handle duplicates

See https://github.com/Automattic/jetpack/pull/25066/files#r919151642

jeherve committed:
Do not include bot comment in comments where we search for refs

See https://github.com/Automattic/jetpack/pull/25066#discussion_r919160659

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -1,63 +1,55 @@
-		response.data.map( comment => {
+	issueComments.map( comment => {
+		if (
+			comment.user.login !== 'github-actions[bot]' &&
+			! comment.body.includes( '**Support References**' )
```
This doesn't properly invert the test above.
```suggestion
			comment.user.login !== 'github-actions[bot]' ||
			! comment.body.includes( '**Support References**' )
```

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -76,32 +68,31 @@ async function getIssueReferences( octokit, owner, repo, number ) {
 		// xxx-zd and xxx-chat, as well as uppercase versions, are considered as alternate versions.
 		const wrongId = supportId.match( /([0-9]*)-(zd|chat)/i );
 		if ( wrongId ) {
 			const correctedId = `${ wrongId[ 1 ] }-${ wrongId[ 2 ] === 'zd' ? 'zen' : 'hc' }`;
```
Should we take the possibility of "ZD" into account?
```suggestion
			const correctedId = `${ wrongId[ 1 ] }-${ wrongId[ 2 ].toLowerCase() === 'zd' ? 'zen' : 'hc' }`;
```

jeherve committed:
Update projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Check previously saved checklist status when updating comment

See https://github.com/Automattic/jetpack/pull/25066#discussion_r919247504

jeherve committed:
Fix match

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -98,38 +100,81 @@ async function createOrUpdateComment( payload, octokit, issueReferences, issueCo
+
+		const {
+			data: { body: existingCommentBody },
+		} = await octokit.rest.issues.getComment( {
```
We already have the body in `issueComments`, no need to hit the API again. Make `getListComment()` return the whole comment instead of the ID instead.

jeherve replied:
👍 Done in cfe41688f84df3123f042f05ff86f3ef01288cde

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -98,38 +100,81 @@ async function createOrUpdateComment( payload, octokit, issueReferences, issueCo
+
+		// First, build a list of all references and whether they are checked or not.
+		const listWithStatusMatch = existingCommentBody.matchAll(
+			/(?:-\s\[(?<checked>\s|x)\]\s)(?<ticketId>[0-9]*-(?:hc|zen))/gm
```
I wonder if we should loosen the ticketid part up to reduce future maintenance burden. Or at least cross-reference this and `referencesRegexP` above with some comments so future-us remember to update both if we add new recognized references.

On the other hand, unless we're worried about people manually editing the bot's comment and screwing up the formatting we could probably get away with `\S+` in here to grab anything on the line.

Also, BTW, you have an unnecessary non-capturing group, and we might want to anchor it.
```suggestion
			/^-\s\[(?<checked>\s|x)\]\s(?<ticketId>\S+)/gm
```

anomiex replied:
Actually, come to think of it, you don't really need the "checked" group. Just match for `/^-\s\[x\]\s(?<ticketId>\S+)/gm` and process all the matches.

jeherve replied:
Good point. e6a12e367df4f4389d5352de1f5dade6f73f20d9 should do it I think.

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -98,38 +100,81 @@ async function createOrUpdateComment( payload, octokit, issueReferences, issueCo
+		const listWithStatusMatch = existingCommentBody.matchAll(
+			/(?:-\s\[(?<checked>\s|x)\]\s)(?<ticketId>[0-9]*-(?:hc|zen))/gm
+		);
+		const statusMatches = [ ...listWithStatusMatch ];
```
No need for this, you can use `listWithStatusMatch` directly in the for-of loop below.

jeherve replied:
Good point. e6a12e367df4f4389d5352de1f5dade6f73f20d9 should do it I think.

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -98,38 +100,81 @@ async function createOrUpdateComment( payload, octokit, issueReferences, issueCo
+		);
+		const statusMatches = [ ...listWithStatusMatch ];
+
+		// Build a new array with only checked ticket references.
+		const checkedList = [];
+		for ( const referenceStatus of statusMatches ) {
+			const [ , checked, ticketId ] = referenceStatus;
+			if ( checked === 'x' ) {
+				checkedList.push( ticketId );
+			}
+		}
+
+		// Remove checked tickets from our existing list of issue references.
+		const uncheckedList = issueReferences.filter(
+			reference => ! checkedList.includes( reference )
+		);
+
+		// Build our comment body, with first the checked references, then the unchecked references.
+		const updatedComment = `**Support References**
+${ checkedList
+	.map(
+		checkedTicket => `
+- [x] ${ checkedTicket }`
+	)
+	.join( '' ) }${ uncheckedList
+			.map(
+				uncheckedTicket => `
+- [ ] ${ uncheckedTicket }`
+			)
+			.join( '' ) }
+
+`;
```
I'd keep the one array of references, and only use `checkedList` to determine whether to `[ ]` or `[x]`. Something like this
```suggestion
		// Extract the checked ticket references.
		const checkedRefs = new Set();
		for ( const referenceStatus of statusMatches ) {
			const [ , checked, ticketId ] = referenceStatus;
			if ( checked === 'x' ) {
				checkedRefs.add( ticketId );
			}
		}

		// Build our comment body, with first the checked references, then the unchecked references.
		const updatedComment = `**Support References**
${ issueReferences
	.map(
		reference => `
- [${ checkedRefs.has( reference ) ? 'x' : ' ' }] ${ reference }`
	)
	.join( '' ) }

`;
```
That way is more likely to keep the references in a stable order, rather than reordering them randomly at some point after someone checks some off and then someone else adds a new comment.

jeherve replied:
It does simplify things a bit, I like it. Done in f5ccd66e34bfe1bb601fd7566540449a3976555e

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,202 @@
+	const supportIds = [];
+	const referencesRegexP = /[0-9]*-(?:chat|hc|zen|zd)/gim;
+
+	debug( `gather-support-references: Getting references from comments.` );
+	issueComments.map( comment => {
+		if (
+			comment.user.login !== 'github-actions[bot]' ||
+			! comment.body.includes( '**Support References**' )
+		) {
+			ticketReferences.push( ...comment.body.matchAll( referencesRegexP ) );
+		}
+	} );
+
+	debug( `gather-support-references: Getting references from issue body.` );
+	const {
+		data: { body },
+	} = await octokit.rest.issues.get( {
+		owner: owner.login,
+		repo,
+		issue_number: +number,
+	} );
+	ticketReferences.push( ...body.matchAll( referencesRegexP ) );
```
Let's swap these two blocks, so references in the description come before references in comments.

Or you could sort the list at some point, if that would make more sense than keeping them in order of first reference.

jeherve replied:
Done in 1e7a941393e664bb1ef8628e8db4206cb6e0c496

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,202 @@
+	} );
+	ticketReferences.push( ...body.matchAll( referencesRegexP ) );
+
+	// Buid a first array with only the support IDs we've collected.
+	ticketReferences.map( reference => {
+		supportIds.push( reference[ 0 ] );
+	} );
+
+	// That array can still include duplicates, or references that are not formatted quite properly.
+	// Let's build a final array with unique and correct support IDs.
+	const correctedSupportIds = new Set();
+	supportIds.map( supportId => {
+		// xxx-zen and xxx-hc are the preferred formats for tickets and chats.
+		// xxx-zd and xxx-chat, as well as uppercase versions, are considered as alternate versions.
+		const wrongId = supportId.match( /([0-9]*)-(zd|chat)/i );
+		if ( wrongId ) {
+			const correctedId = `${ wrongId[ 1 ] }-${
+				wrongId[ 2 ].toLowerCase() === 'zd' ? 'zen' : 'hc'
+			}`;
+			correctedSupportIds.add( correctedId );
+		} else {
+			correctedSupportIds.add( supportId.toLowerCase() );
+		}
+	} );
```
These two blocks could be merged, process `ticketReferences` into `correctedSupportIds` directly without the intermediate step of `supportIds`.

jeherve replied:
Done in ccc8028e6588f19a217a95326578e7acfce2d855

anomiex commented on projects/github-actions/repo-gardening/src/tasks/gather-support-references/index.js:
```diff
@@ -0,0 +1,202 @@
+		// If no comment was published before, publish one now.
+		debug( `gather-support-references: Posting comment to issue #${ number }` );
+
+		const comment = `**Support References**
+${ issueReferences
+	.map(
+		reference => `
+- [ ] ${ reference }`
+	)
+	.join( '' ) }
+
+`;
```
It's be nice not to mostly-duplicate this bit from above. You could do a structure like
```
const checkedRefs = new Set();
if ( existingComment ) {
    populate checkedRefs
}

const comment = ...;
if ( existingComment ) {
    update
} else {
    create
]
```
Or you could just make a `buildComment( issueReferences, checkedRefs )` function to call.

jeherve replied:
Done in eb6572b119d8e5291559910a99bedd27e19f70db

jeherve committed:
Use previously fetched comment info instead of fetching twice

See https://github.com/Automattic/jetpack/pull/25066#discussion_r921143428

jeherve committed:
Simplify extraction of checked tickets

See:
- https://github.com/Automattic/jetpack/pull/25066/files#r921181874
- https://github.com/Automattic/jetpack/pull/25066/files#r921158405

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Add references from original issue body first

See https://github.com/Automattic/jetpack/pull/25066/files#r921168655

jeherve committed:
Simplify extracting correct support IDs from content

See https://github.com/Automattic/jetpack/pull/25066/files#r921171150

jeherve committed:
Build updated comment issue list from a single array

See https://github.com/Automattic/jetpack/pull/25066#discussion_r921166492

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Extract comment string building logic

See https://github.com/Automattic/jetpack/pull/25066#discussion_r921177160

jeherve commented:
@anomiex Thanks for all the super useful feedback, as always :) This should be good for another review now.

anomiex reviewed (approved):
Let's give it a try!
